src/main/kotlin/grpc/AuthenticationInterceptor.kt

```diff
@@ -0,0 +1,298 @@
+    }
+}
+
+/**
+ * A set of public gRPC methods that do not require authentication.
+ */
+private val PUBLIC_METHODS =
+    setOf(
+        "snowballr.SnowballR/GetAuthenticationStatus",
+        "snowballr.SnowballR/Register",
+        "snowballr.SnowballR/Login",
+        "snowballr.SnowballR/RequestPasswordReset",
+        "snowballr.SnowballR/ResetPassword",
+    )
+
+/**
+ * Parses a "Cookie" header string into a map of key-value pairs.
+ * @param cookieHeader The string value of the "Cookie" header.
+ * @return A map of cookie names to cookie values.
+ */
+private fun parseCookies(cookieHeader: String?): Map<String, String> {
+    if (cookieHeader.isNullOrBlank()) return emptyMap()
+    return cookieHeader
+        .split(";")
+        .map { it.trim() }
+        .filter { it.contains('=') }
+        .associate {
+            val parts = it.split('=', limit = 2)
+            parts[0] to parts[1]
+        }
+}
+
+/**
+ * A [ServerInterceptor] for handling authentication via JWT.
+ *
+ * This interceptor performs the following steps:
+ * 1. Checks if the called method is public (e.g., login, register) or a health check.
+ *    If so, it bypasses authentication.
+ * 2. For protected methods, it extracts the access and refresh tokens from the gRPC metadata.
+ * 3. It validates the access token. If valid, it extracts the user ID, attaches it to the gRPC
+ *    [Context], and allows the call to proceed.
+ * 4. If the access token is invalid (e.g., expired), it attempts to use the refresh token to
+ *    generate a new access token.
+ * 5. If a new access token is successfully generated, it's sent back to the client in the
+ *    response trailers, and the call proceeds with the user ID attached to the context.
+ * 6. If authentication fails at any point (e.g., missing tokens, invalid refresh token),
+ *    the call is rejected with a [Status.UNAUTHENTICATED] status.
+ */
+val authenticationInterceptor =
+    object : ServerInterceptor {
+        /**
+         * Returns a no-op listener. Used when the call chain cannot proceed.
+         */
+        private fun <ReqT> emptyListener(): ServerCall.Listener<ReqT?> = object : ServerCall.Listener<ReqT?>() {}
+
+        override fun <ReqT : Any?, RespT : Any?> interceptCall(
+            call: ServerCall<ReqT?, RespT?>?,
+            headers: Metadata?,
+            next: ServerCallHandler<ReqT?, RespT?>?,
+        ): ServerCall.Listener<ReqT?>? {
+            val methodName = call?.methodDescriptor?.fullMethodName
+            logger.debug { "Auth Interceptor: Intercepting call to $methodName" }
+
+            if (methodName == null) {
+                call?.close(Status.UNAUTHENTICATED.withDescription("Method name is null"), Metadata())
+                return emptyListener()
+            }
+
+            // Bypass authentication for public methods / health checks
+            if (methodName in PUBLIC_METHODS || methodName.startsWith("grpc.health.v1.Health")) {
```

Comment:
Same as for the other methods, we can use a constant here.
```suggestion
            if (methodName in PUBLIC_METHODS || methodName.startsWith(HealthGrpc.SERVICE_NAME)) {
```